prometheus/metric.go

```diff
@@ -183,9 +184,16 @@ func (m *withExemplarsMetric) Write(pb *dto.Metric) error {
 			})
 			if i < len(pb.Histogram.Bucket) {
 				pb.Histogram.Bucket[i].Exemplar = e
-			} else {
-				// This is not possible as last bucket is Inf.
-				panic("no bucket was found for given exemplar value")
+			} else { // +inf bucket should be explicitly added if there is an exemplar for it.
+				b := &dto.Bucket{
+					CumulativeCount: proto.Uint64(pb.Histogram.Bucket[len(pb.Histogram.GetBucket())-1].GetCumulativeCount()),
+					UpperBound:      proto.Float64(math.Inf(1)),
+					Exemplar:        e,
+				}
+				pb.Histogram.Bucket = append(pb.Histogram.Bucket, b)
+				break
+				// end looping after creating +inf bucket and adding one exemplar.
+				// there could be other exemplars that are in the "inf" range but those will be ignored.
```

Comment:
If we want those comments, let's make them full sentence, but also I am not sure this is safe. Nowhere in the interface/signature we mention that exemplars will be sorted by anything. I think there is no harm to continue the loop, unless we want to optimize this some day. WDYT?

```suggestion
```

Reply:
Yes you are right that there is no mention (or expectation) of sorting exemplars. If we leave the loop running we pick the last exemplar and if we leave it as is and terminate it here, we pick the first exemplar (in that range) - sort order of the exemplar value being arbitrary.  I would consider terminating the loop as it would just avoid running through the remaining exemplars that we are not going to use anyway. But please let me know if you think otherwise or see any safety concerns, I can make the change. 

Also, I made the comment into a sentence.

Reply:
This was not addressed completely. What do you about lack of sorted order invariant?

> Nowhere in the interface/signature we mention that exemplars will be sorted by anything. I think there is no harm to continue the loop, unless we want to optimize this some day. WDYT?

Reply:
For all the buckets we are comparing the exemplar values to the bucket bound and matching the right exemplar to right bucket (this is the exiting logic): 

specifically the check here to get the index of the right bucket for that exemplar: 

https://github.com/prometheus/client_golang/blob/a528affed905f64ed33ed2f52b7880ed7247f99e/prometheus/metric.go#L182

And assigning the exemplar to the right bucket here:

https://github.com/prometheus/client_golang/blob/a528affed905f64ed33ed2f52b7880ed7247f99e/prometheus/metric.go#L185
 

At a high level the only change in this PR is that instead of a panic in the [else](https://github.com/prometheus/client_golang/blob/a528affed905f64ed33ed2f52b7880ed7247f99e/prometheus/metric.go#L186) condition, we add the `+Inf` bucket and add one exemplar that is outside of all previous bucket range and break the loop.

If there are multiple exemplars for the `+Inf` bucket, we could pick the exemplar that is more representative of the group such as a median - a future improvement, would require further discussion. Currently we are just picking the first in the array in the +inf bucket range.

I hope that clarifies your concern, if I understood your question correctly.

Reply:
@bwplotka waiting for your response :) Would be nice if we could include this in the upcoming release.

Reply:
Missed this discussion, sorry. 

So I like the idea about median or tuning the exemplar we take from those belonging to +Inf. My only problem is that those inputs can be not sorted, that's it. Hope this PR https://github.com/prometheus/client_golang/pull/1100 makes sense to you.